ltm/agent.py

```diff
@@ -1,102 +1,408 @@
+        while True:
+            try:
+                print("Keyword gen")
+                response = ask_llm(context, model=self.model, max_overall_tokens=self.max_prompt_size, cost_callback=cost_cb, temperature=self.temperature)
+
+                keywords = [k.lower() for k in sanitize_and_parse_json(response)]
+                break
+            except Exception as e:
+                print(repr(e) + response)
+                continue
+
+        # Update known list of keywords
+        for k in keywords:
+            if k not in self.defined_kws:
+                self.defined_kws.append(k)
+
+        print(f"Interaction keywords: {keywords}")
+        return keywords
+
+    def create_context(self, user_message, max_prompt_size, previous_interactions, cost_cb):
+
+        stamped_user_message = str(self.now) + ": " + user_message
+        context = [make_system_message(self.system_message), make_user_message(stamped_user_message)]
+        relevant_interactions = self.get_relevant_memories(user_message, cost_cb)
+
+        # Get interactions from the memories
+        for m in relevant_interactions:
+            if "trivia" in m[0].content:
+                colour_print("YELLOW", f"<*** trivia ***>")
+            else:
+                colour_print("YELLOW", f"{m[0].content}")
+
+        # Add the previous messages
+        final_idx = self.session.message_count - 1
+        while previous_interactions > 0 and final_idx > 0:
+
+            # Agent reply
+            context.insert(1, self.session.by_index(final_idx).as_llm_dict())
+            # User message
+            context.insert(1, self.session.by_index(final_idx-1).as_llm_dict())
+
+            final_idx -= 2
+            previous_interactions -= 1
+
+        # Add in memories up to the max prompt size
+        current_size = token_counter(self.model, messages=context)
+        shown_mems = 0
+        target_size = max_prompt_size - self.max_message_size
+
+        for interaction in relevant_interactions[::-1]:
+            user_interaction, assistant_interaction = interaction
+            future_size = token_counter(self.model, messages=context + [user_interaction.as_llm_dict(), assistant_interaction.as_llm_dict()])
+
+            # If this message is going to be too big, then skip it
+            if shown_mems >= 100:
+                break
+
+            if future_size > target_size:
+                continue
+
+            # Add the interaction and count the tokens
+            context.insert(1, assistant_interaction.as_llm_dict())
+
+            ts = datetime.datetime.fromtimestamp(user_interaction.timestamp)
+            et_descriptor = f"{str(ts)[:-7]} ({td_format(self.now - ts)}) "
+            context.insert(1, user_interaction.as_llm_dict())
+            context[1]["content"] = et_descriptor + context[1]["content"]
+
+            shown_mems += 1
+
+            current_size = future_size
```

Comment:
Just a couple of minor optimizations:
1. I find `reversed(relevant_interactions)` more readable, and it also does not make a copy of the list (`[::-1]` does).
2. `if show_mems >= 100` could go right after increasing the counter in line 197.